File: leetcode/198. House Robber/main.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int maxMoney = 1;
        
        int n = nums.size();
        
        if(n == 1) return nums[0];
        else if(n == 2) return max(nums[0], nums[1]);
        
        int lookup[nums.size()];
        
        // vector<int> lookup;
        // for(int i=0; i<n; i++) lookup.push_back(-1);
        lookup[0] = nums[0];
        lookup[1] = max(nums[0], nums[1]);
        
        for(int i=3; i<=n; i++){
            lookup[i-1] = max(lookup[i-3] + nums[i-1], lookup[i-2]);
        }
        
        return lookup[n-1];
    }
};

File: leetcode/198. House Robber/main.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        # dp[n-1]: max money can be robbed in n houses
        # memoization approach
        
        n = len(nums)
        dp = [-1 for i in range(n)]
        dp[0] = nums[0]
        
        if n > 1:
            dp[1] = max(nums[0], nums[1])
        
        def find_sol(nums, dp, n):
            if dp[n] == -1:
                dp[n] = max(find_sol(nums, dp, n-1), find_sol(nums, dp, n-2) + nums[n])
            return dp[n]
        
        return find_sol(nums, dp, n-1)
        